akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ShardedDaemonProcessKeepAlivePinger.scala

```diff
@@ -0,0 +1,292 @@
+
+        waitForInitialDdataState()
+      }
+    }
+  }
+
+  private def waitForInitialDdataState(): Behavior[Message] = {
+
+    Behaviors.receiveMessagePartial {
+      case InternalGetResponse(success @ Replicator.GetSuccess(`ddataKey`)) =>
+        val state = success.get(ddataKey).value
+        subscribeToTopic()
+        if (state.completed) {
+          context.log.debugN(
+            "{}: Starting pinger with [{}] workers, revision [{}] from [{}]",
+            daemonProcessName,
+            state.numberOfProcesses,
+            state.revision,
+            state.started)
+          startTicks(state.revision)
+          start(state.revision, state.numberOfProcesses)
+        } else {
+          context.log.debugN(
+            "{}: Pinger found non completed state with [{}] workers, revision [{}] from [{}]",
+            daemonProcessName,
+            state.revision,
+            state.started)
+          // FIXME what about ticks here?
+          // FIXME is pausing the right thing?
+          paused(state.revision - 1)
+        }
+
+      case InternalGetResponse(_ @Replicator.NotFound(`ddataKey`)) =>
+        context.log.debug(
+          "{}: No previous state stored for Sharded Daemon Process, starting from revision [0] with the initial [{}] workers",
+          daemonProcessName,
+          initialNumberOfInstances)
+        subscribeToTopic()
+        startTicks(startRevision)
+        start(startRevision, numberOfProcesses = initialNumberOfInstances)
+
+      case InternalGetResponse(_ @Replicator.GetFailure(`ddataKey`)) =>
+        context.log
+          .debug("{}: Failed fetching initial state for Sharded Daemon Process pinger, retrying", daemonProcessName)
+        init()
+
+    }
+
+  }
+
+  private def start(currentRevision: Long, numberOfProcesses: Int): Behavior[Message] = {
+    val sortedIdentities =
+      ShardedDaemonProcessId.sortedIdentitiesFor(currentRevision, numberOfProcesses, supportsRescale)
+
+    Behaviors.receiveMessage {
+      case Tick(`currentRevision`) =>
+        if (shouldPing()) {
+          context.log.debugN(
+            s"Sending periodic keep alive for Sharded Daemon Process [{}] to [{}] processes (revision [{}]).",
+            daemonProcessName,
+            sortedIdentities.size,
+            currentRevision)
+          context.pipeToSelf(sendKeepAliveMessages(sortedIdentities)) { _ =>
+            SendKeepAliveDone(currentRevision)
+          }
+        } else {
+          timers.startSingleTimer(Tick, Tick(currentRevision), settings.keepAliveInterval)
+        }
+        Behaviors.same
+      case SendKeepAliveDone(`currentRevision`) =>
+        timers.startSingleTimer(Tick, Tick(currentRevision), settings.keepAliveInterval)
+        Behaviors.same
+      case Pause(revision, replyTo) =>
+        context.log.debug2("Pausing sharded daemon process pinger [{}] (revision [{}])", daemonProcessName, revision)
+        if (revision <= currentRevision) {
```

Comment:
Not sure I understand this, and why the `revision` is included in the `Tick(revision)`? Can't it just always cancel when switching to paused. I see that there are some Tick that are also sent to self without timer. Maybe the revision is needed because of that but feels a little more complicated than necessary?

Reply:
Leftover from previous iteration I think, removed/back to case object in e5f6d4e11a5073539ff16d6ad65a1cc33c4f2b5d